File: eve_mcp/names.py
"""Turn ESI's bare numeric ids into names the LLM can actually reason about."""
from __future__ import annotations

import asyncio
import logging
import time
from collections import defaultdict
from dataclasses import dataclass
from typing import Any, Iterable, Sequence

from .cache import Store
from .config import JITA_4_4_STATION_ID, THE_FORGE_REGION_ID
from .esi import EsiClient, EsiError

log = logging.getLogger("eve_mcp.names")

_NAME_BATCH = 900
#: /universe/ids rejects duplicates and caps a request at 500 names.
_IDS_BATCH = 500
#: Location ids at or above this are player structures, which /universe/names
#: refuses to resolve; they need the authenticated structure endpoint instead.
#: Upwell structures are item ids and start at 1e12 — everything below that,
#: characters (2.1e9) very much included, belongs to /universe/names.
_STRUCTURE_ID_FLOOR = 1_000_000_000_000
#: Market prices change slowly enough that an hour-old snapshot is fine.
_PRICE_TTL = 3600.0

#: Categories POST /universe/ids answers with, mapped to the label used in
#: error strings and confirmation previews.
ID_CATEGORIES: dict[str, str] = {
    "agents": "agent",
    "alliances": "alliance",
    "characters": "character",
    "constellations": "constellation",
    "corporations": "corporation",
    "factions": "faction",
    "inventory_types": "item type",
    "regions": "region",
    "stations": "station",
    "systems": "solar system",
}


def _article(kind: str) -> str:
    return f"{'an' if kind[:1].lower() in 'aeiou' else 'a'} {kind}"


@dataclass(frozen=True)
class NameMatch:
    id: int
    name: str
    #: The /universe/ids key this came from, e.g. "inventory_types".
    category: str
    #: Human label for that category, e.g. "item type".
    kind: str


@dataclass(frozen=True)
class NameResolution:
    query: str
    chosen: NameMatch | None
    alternatives: tuple[NameMatch, ...] = ()

    @property
    def ambiguous(self) -> bool:
        return bool(self.alternatives)

    def describe(self) -> str:
        """'Rifter' is an item type (#587) and also a character (#187399875)."""
        if self.chosen is None:
            return f"{self.query!r} matched nothing"
        others = ", ".join(f"{_article(m.kind)} (#{m.id})" for m in self.alternatives)
        chosen = f"{_article(self.chosen.kind)} (#{self.chosen.id})"
        return f"{self.query!r} is {chosen} and also {others}"


class Resolver:
    def __init__(self, esi: EsiClient, store: Store) -> None:
        self._esi = esi
        self._store = store
        self._prices: dict[int, dict[str, float]] | None = None
        self._prices_at = 0.0
        self._price_lock = asyncio.Lock()

    # ------------------------------------------------------------------ names

    async def names(self, ids: Iterable[int], character_id: int | None = None) -> dict[int, str]:
        """Resolve any mix of character/corp/type/system/station ids to names."""
        wanted = {int(i) for i in ids if i}
        if not wanted:
            return {}

        cached = await self._store.get_names(list(wanted))
        out = {i: v["name"] for i, v in cached.items()}
        missing = sorted(wanted - set(out))
        if not missing:
            return out

        universal = [i for i in missing if i < _STRUCTURE_ID_FLOOR]
        structures = [i for i in missing if i >= _STRUCTURE_ID_FLOOR]

        for start in range(0, len(universal), _NAME_BATCH):
            chunk = universal[start : start + _NAME_BATCH]
            try:
                result = await self._esi.post("/universe/names", json_body=chunk)
            except EsiError as exc:
                # A single unresolvable id makes the whole batch 404; degrade quietly.
                log.info("bulk name lookup failed for %d ids: %s", len(chunk), exc)
                continue
            entries = [
                (int(row["id"]), row["name"], row.get("category"))
                for row in result or []
                if row.get("id") and row.get("name")
            ]
            await self._store.put_names(entries)
            out.update({i: n for i, n, _ in entries})

        if structures and character_id is not None:
            resolved = await asyncio.gather(
                *(self._structure_name(sid, character_id) for sid in structures),
                return_exceptions=True,
            )
            entries = []
            for sid, name in zip(structures, resolved):
                if isinstance(name, str):
                    entries.append((sid, name, "structure"))
                    out[sid] = name
            await self._store.put_names(entries)

        for missing_id in wanted - set(out):
            out[missing_id] = f"Unknown #{missing_id}"
        return out

    async def name(self, id_: int, character_id: int | None = None) -> str:
        return (await self.names([id_], character_id)).get(id_, f"Unknown #{id_}")

    async def _structure_name(self, structure_id: int, character_id: int) -> str:
        result = await self._esi.get(
            f"/universe/structures/{structure_id}", character_id=character_id
        )
        return result.data.get("name", f"Structure #{structure_id}")

    async def ids_from_names(self, names: Sequence[str]) -> dict[str, Any]:
        """Reverse lookup: exact names -> ids, grouped by category.

        ESI rejects the whole request with a 400 when the list holds duplicates
        or more than 500 entries, and callers legitimately produce both — a fit
        with two identical guns, a route whose `avoid` repeats the origin. Both
        are handled here so no call site has to remember.
        """
        unique = list(dict.fromkeys(n.strip() for n in names if n and n.strip()))
        if not unique:
            return {}
        out: dict[str, Any] = {}
        for start in range(0, len(unique), _IDS_BATCH):
            part = await self._esi.post(
                "/universe/ids", json_body=unique[start : start + _IDS_BATCH]
            )
            for key, rows in (part or {}).items():
                out.setdefault(key, []).extend(rows or [])
        return out

    async def resolve_names(
        self,
        names: Sequence[str],
        *,
        prefer: Sequence[str] = (),
        only: Sequence[str] = (),
    ) -> dict[str, NameResolution]:
        """Resolve exact names to one entity each, keyed by lower-cased query.

        `/universe/ids` answers grouped by category, and one string can match in
        several at once — 'Rifter' is both a player character and a ship type.
        Taking entries[0] of a category therefore picks a coin-flip winner, and
        looping every category picks several. Regrouping by name turns that into
        a value the caller has to deal with.

        `only` restricts matching to those categories: an `item` parameter can
        only ever mean an inventory type, so a same-named character is not an
        ambiguity there. `prefer` orders them when several still match; unlisted
        categories sort last, then by category name, then by ascending id — so
        the choice is always deterministic.
        """
        lookup = await self.ids_from_names(names)
        buckets: dict[str, list[NameMatch]] = defaultdict(list)
        for key, entries in (lookup or {}).items():
            if only and key not in only:
                continue
            kind = ID_CATEGORIES.get(key, key)
            for entry in entries or []:
                if not entry.get("id") or not entry.get("name"):
                    continue
                buckets[entry["name"].strip().lower()].append(
                    NameMatch(int(entry["id"]), entry["name"], key, kind)
                )

        rank = {key: i for i, key in enumerate(prefer)}
        out: dict[str, NameResolution] = {}
        for asked in names:
            wanted = (asked or "").strip().lower()
            matches = sorted(
                buckets.get(wanted, []),
                key=lambda m: (rank.get(m.category, len(rank)), m.category, m.id),
            )
            out[wanted] = NameResolution(
                query=(asked or "").strip(),
                chosen=matches[0] if matches else None,
                alternatives=tuple(matches[1:]),
            )
        return out

    # ------------------------------------------------------------------ types

    async def type_info(self, type_id: int) -> dict[str, Any]:
        key = f"type:{type_id}"
        cached = await self._store.get_blob(key, max_age=30 * 86400)
        if cached is not None:
            return cached
        result = await self._esi.get(f"/universe/types/{type_id}")
        await self._store.put_blob(key, result.data)
        return result.data

    async def group_name(self, group_id: int) -> str:
        """Groups have their own id space, separate from /universe/names."""
        if not group_id:
            return "unknown"
        key = f"group:{group_id}"
        cached = await self._store.get_blob(key, max_age=30 * 86400)
        if cached is None:
            try:
                result = await self._esi.get(f"/universe/groups/{group_id}")
            except EsiError:
                return f"Group #{group_id}"
            cached = result.data
            await self._store.put_blob(key, cached)
        return (cached or {}).get("name", f"Group #{group_id}")

    async def type_infos(self, type_ids: Iterable[int]) -> dict[int, dict[str, Any]]:
        unique = sorted({int(t) for t in type_ids if t})
        results = await asyncio.gather(
            *(self.type_info(t) for t in unique), return_exceptions=True
        )
        out: dict[int, dict[str, Any]] = {}
        for type_id, info in zip(unique, results):
            if isinstance(info, dict):
                out[type_id] = info
        return out

    # ----------------------------------------------------------------- prices

    async def reference_prices(self) -> dict[int, dict[str, float]]:
        """CCP's global average/adjusted price per type — the appraisal default."""
        async with self._price_lock:
            # The in-memory copy needs the same TTL as the blob, or the process
            # pins the first snapshot it ever fetched for as long as it runs.
            if self._prices is not None and time.time() - self._prices_at < _PRICE_TTL:
                return self._prices
            cached = await self._store.get_blob("markets:prices", max_age=_PRICE_TTL)
            if cached is None:
                result = await self._esi.get("/markets/prices")
                cached = {
                    str(row["type_id"]): {
                        "average": float(row.get("average_price") or 0.0),
                        "adjusted": float(row.get("adjusted_price") or 0.0),
                    }
                    for row in result.data or []
                }
                await self._store.put_blob("markets:prices", cached)
            self._prices = {int(k): v for k, v in cached.items()}
            self._prices_at = time.time()
            return self._prices

    async def reference_price(self, type_id: int) -> float:
        prices = await self.reference_prices()
        entry = prices.get(int(type_id), {})
        return entry.get("average") or entry.get("adjusted") or 0.0

    async def hub_quotes(
        self,
        type_id: int,
        region_id: int = THE_FORGE_REGION_ID,
        station_id: int | None = JITA_4_4_STATION_ID,
    ) -> dict[str, Any]:
        """Best buy/sell prices for one type at a trade hub, from live orders."""
        result = await self._esi.get_all_pages(
            f"/markets/{region_id}/orders",
            params={"type_id": type_id, "order_type": "all"},
            max_pages=10,
        )
        orders = [o for o in (result.data or []) if isinstance(o, dict)]
        if station_id is not None:
            at_hub = [o for o in orders if o.get("location_id") == station_id]
            if at_hub:
                orders = at_hub
        buys = sorted(
            (o["price"] for o in orders if o.get("is_buy_order")), reverse=True
        )
        sell_prices = sorted(o["price"] for o in orders if not o.get("is_buy_order"))
        return {
            "type_id": type_id,
            "region_id": region_id,
            "station_id": station_id,
            "best_sell": sell_prices[0] if sell_prices else None,
            "best_buy": buys[0] if buys else None,
            "sell_order_count": len(sell_prices),
            "buy_order_count": len(buys),
            "sell_volume": sum(
                o.get("volume_remain", 0) for o in orders if not o.get("is_buy_order")
            ),
            "buy_volume": sum(
                o.get("volume_remain", 0) for o in orders if o.get("is_buy_order")
            ),
            "data_age": result.stale_note,
        }

File: internal/adapter/names/names.go
package names

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/truewebber/eve-online-mcp/internal/adapter/esi"
	"github.com/truewebber/eve-online-mcp/internal/adapter/store"
	"github.com/truewebber/eve-online-mcp/internal/domain/j"
)

const (
	nameBatch        = 900
	idsBatch         = 500
	structureIDFloor = 1_000_000_000_000
	priceTTL         = 3600.0
	staticBlobAge    = 30 * 24 * time.Hour
	hubOrderPages    = 10
)

func categoryKind(key string) string {
	switch key {
	case "agents":
		return "agent"
	case "alliances":
		return "alliance"
	case "characters":
		return "character"
	case "constellations":
		return "constellation"
	case "corporations":
		return "corporation"
	case "factions":
		return "faction"
	case "inventory_types":
		return "item type"
	case "regions":
		return "region"
	case "stations":
		return "station"
	case "systems":
		return "solar system"
	default:
		return ""
	}
}

type NameMatch struct {
	ID       int
	Name     string
	Category string
	Kind     string
}

type NameResolution struct {
	Query        string
	Chosen       *NameMatch
	Alternatives []NameMatch
}

func (r NameResolution) Ambiguous() bool { return len(r.Alternatives) > 0 }

func (r NameResolution) Describe() string {
	if r.Chosen == nil {
		return fmt.Sprintf("%q matched nothing", r.Query)
	}
	others := make([]string, 0, len(r.Alternatives))
	for _, m := range r.Alternatives {
		others = append(others, article(m.Kind)+fmt.Sprintf(" (#%d)", m.ID))
	}
	chosen := article(r.Chosen.Kind) + fmt.Sprintf(" (#%d)", r.Chosen.ID)

	return fmt.Sprintf("%q is %s and also %s", r.Query, chosen, strings.Join(others, ", "))
}

func article(kind string) string {
	if kind == "" {
		return "a thing"
	}
	switch strings.ToLower(kind[:1]) {
	case "a", "e", "i", "o", "u":
		return "an " + kind
	}

	return "a " + kind
}

type Resolver struct {
	esi      *esi.Client
	store    *store.Store
	prices   map[int]map[string]float64
	pricesAt time.Time
	priceMu  sync.Mutex
}

func New(e *esi.Client, db *store.Store) *Resolver {
	return &Resolver{esi: e, store: db}
}

func (r *Resolver) Names(ctx context.Context, ids []int, characterID *int) (map[int]string, error) {
	wanted := uniquePositive(ids)
	if len(wanted) == 0 {
		return map[int]string{}, nil
	}
	out, missing, err := r.lookupCachedNames(ctx, wanted)
	if err != nil {
		return nil, err
	}
	if len(missing) == 0 {
		return out, nil
	}
	r.fillMissingNames(ctx, out, missing, characterID)
	for id := range wanted {
		if _, ok := out[id]; !ok {
			out[id] = fmt.Sprintf("Unknown #%d", id)
		}
	}

	return out, nil
}

func (r *Resolver) Name(ctx context.Context, id int, characterID *int) (string, error) {
	m, err := r.Names(ctx, []int{id}, characterID)
	if err != nil {
		return fmt.Sprintf("Unknown #%d", id), err
	}
	if n, ok := m[id]; ok {
		return n, nil
	}

	return fmt.Sprintf("Unknown #%d", id), nil
}

func (r *Resolver) IDsFromNames(ctx context.Context, names []string) (map[string]any, error) {
	seen := map[string]struct{}{}
	var unique []string
	for _, n := range names {
		n = strings.TrimSpace(n)
		if n == "" {
			continue
		}
		if _, ok := seen[n]; ok {
			continue
		}
		seen[n] = struct{}{}
		unique = append(unique, n)
	}
	if len(unique) == 0 {
		return map[string]any{}, nil
	}
	out := map[string]any{}
	for start := 0; start < len(unique); start += idsBatch {
		end := min(start+idsBatch, len(unique))
		part, err := r.esi.Post(ctx, "/universe/ids", nil, nil, unique[start:end])
		if err != nil {
			return nil, err
		}
		for key, rows := range j.Map(part) {
			existing := j.Slice(out[key])
			out[key] = append(existing, j.Slice(rows)...)
		}
	}

	return out, nil
}

func (r *Resolver) ResolveNames(ctx context.Context, names []string, prefer, only []string) (map[string]NameResolution, error) {
	lookup, err := r.IDsFromNames(ctx, names)
	if err != nil {
		return nil, err
	}

	return pickNameResolutions(names, collectNameBuckets(lookup, only), preferRank(prefer)), nil
}

func (r *Resolver) TypeInfo(ctx context.Context, typeID int) (map[string]any, error) {
	key := fmt.Sprintf("type:%d", typeID)
	maxAge := staticBlobAge
	cached, err := r.blob(ctx, key, &maxAge)
	if err != nil {
		return nil, err
	}
	if cached != nil {
		return j.Map(cached), nil
	}
	result, err := r.esi.Get(ctx, fmt.Sprintf("/universe/types/%d", typeID), nil, nil, nil)
	if err != nil {
		return nil, err
	}
	data := j.Map(result.Data)
	if err := r.putBlob(ctx, key, data); err != nil {
		log.Printf("names: put type %s: %v", key, err)
	}

	return data, nil
}

func (r *Resolver) GroupName(ctx context.Context, groupID int) string {
	if groupID == 0 {
		return "unknown"
	}
	key := fmt.Sprintf("group:%d", groupID)
	maxAge := staticBlobAge
	cached, err := r.blob(ctx, key, &maxAge)
	if err != nil || cached == nil {
		result, err := r.esi.Get(ctx, fmt.Sprintf("/universe/groups/%d", groupID), nil, nil, nil)
		if err != nil {
			return fmt.Sprintf("Group #%d", groupID)
		}
		cached = result.Data
		if err := r.putBlob(ctx, key, cached); err != nil {
			log.Printf("names: put group %s: %v", key, err)
		}
	}
	name := j.Str(j.Map(cached)["name"])
	if name == "" {
		return fmt.Sprintf("Group #%d", groupID)
	}

	return name
}

func (r *Resolver) TypeInfos(ctx context.Context, typeIDs []int) map[int]map[string]any {
	seen := map[int]struct{}{}
	var unique []int
	for _, t := range typeIDs {
		if t == 0 {
			continue
		}
		if _, ok := seen[t]; ok {
			continue
		}
		seen[t] = struct{}{}
		unique = append(unique, t)
	}
	out := map[int]map[string]any{}
	type box struct {
		id   int
		info map[string]any
	}
	ch := make(chan box, len(unique))
	for _, t := range unique {
		go func(t int) {
			info, err := r.TypeInfo(ctx, t)
			if err == nil {
				ch <- box{t, info}
			} else {
				ch <- box{t, nil}
			}
		}(t)
	}
	for range unique {
		b := <-ch
		if b.info != nil {
			out[b.id] = b.info
		}
	}

	return out
}

func (r *Resolver) ReferencePrices(ctx context.Context) (map[int]map[string]float64, error) {
	r.priceMu.Lock()
	defer r.priceMu.Unlock()
	if r.prices != nil && time.Since(r.pricesAt) < time.Duration(priceTTL*float64(time.Second)) {
		return r.prices, nil
	}
	ttl := time.Duration(priceTTL * float64(time.Second))
	cached, err := r.blob(ctx, "markets:prices", &ttl)
	if err != nil {
		return nil, err
	}
	if cached == nil {
		result, err := r.esi.Get(ctx, "/markets/prices", nil, nil, nil)
		if err != nil {
			return nil, err
		}
		blob := map[string]any{}
		for _, row := range j.Maps(result.Data) {
			blob[strconv.Itoa(j.Int(row["type_id"]))] = map[string]any{
				"average":  j.Float(row["average_price"]),
				"adjusted": j.Float(row["adjusted_price"]),
			}
		}
		if err := r.putBlob(ctx, "markets:prices", blob); err != nil {
			log.Printf("names: put market prices: %v", err)
		}
		cached = blob
	}
	prices := map[int]map[string]float64{}
	for k, v := range j.Map(cached) {
		var id int
		if _, err := fmt.Sscanf(k, "%d", &id); err != nil {
			continue
		}
		m := j.Map(v)
		prices[id] = map[string]float64{"average": j.Float(m["average"]), "adjusted": j.Float(m["adjusted"])}
	}
	r.prices = prices
	r.pricesAt = time.Now()

	return prices, nil
}

func (r *Resolver) ReferencePrice(ctx context.Context, typeID int) float64 {
	prices, err := r.ReferencePrices(ctx)
	if err != nil {
		return 0
	}
	entry := prices[typeID]
	if entry["average"] != 0 {
		return entry["average"]
	}

	return entry["adjusted"]
}

func (r *Resolver) HubQuotes(ctx context.Context, typeID, regionID int, stationID *int) (map[string]any, error) {
	result, err := r.esi.GetAllPages(
		ctx,
		fmt.Sprintf("/markets/%d/orders", regionID),
		nil,
		map[string]any{"type_id": typeID, "order_type": "all"},
		hubOrderPages,
	)
	if err != nil {
		return nil, err
	}
	orders := j.Maps(result.Data)
	if stationID != nil {
		var atHub []map[string]any
		for _, o := range orders {
			if j.Int(o["location_id"]) == *stationID {
				atHub = append(atHub, o)
			}
		}
		if len(atHub) > 0 {
			orders = atHub
		}
	}
	var buys, sells []float64
	var sellVol, buyVol float64
	for _, o := range orders {
		price := j.Float(o["price"])
		vol := j.Float(o["volume_remain"])
		if j.Bool(o["is_buy_order"]) {
			buys = append(buys, price)
			buyVol += vol
		} else {
			sells = append(sells, price)
			sellVol += vol
		}
	}
	sort.Float64s(sells)
	sort.Sort(sort.Reverse(sort.Float64Slice(buys)))
	out := map[string]any{
		"type_id": typeID, "region_id": regionID,
		"best_sell": nil, "best_buy": nil,
		"sell_order_count": len(sells), "buy_order_count": len(buys),
		"sell_volume": sellVol, "buy_volume": buyVol,
		"data_age": result.StaleNote(),
	}
	if stationID != nil {
		out["station_id"] = *stationID
	}
	if len(sells) > 0 {
		out["best_sell"] = sells[0]
	}
	if len(buys) > 0 {
		out["best_buy"] = buys[0]
	}

	return out, nil
}

func uniquePositive(ids []int) map[int]struct{} {
	wanted := map[int]struct{}{}
	for _, id := range ids {
		if id != 0 {
			wanted[id] = struct{}{}
		}
	}

	return wanted
}

func (r *Resolver) lookupCachedNames(ctx context.Context, wanted map[int]struct{}) (map[int]string, []int, error) {
	list := make([]int, 0, len(wanted))
	for id := range wanted {
		list = append(list, id)
	}
	id64s := make([]int64, len(list))
	for i, id := range list {
		id64s[i] = int64(id)
	}
	cached, err := r.store.NameGet(ctx, id64s)
	if err != nil {
		return nil, nil, err
	}
	out := map[int]string{}
	var missing []int
	for id := range wanted {
		if row, ok := cached[int64(id)]; ok {
			out[id] = row.Name
		} else {
			missing = append(missing, id)
		}
	}

	return out, missing, nil
}

func (r *Resolver) fillMissingNames(ctx context.Context, out map[int]string, missing []int, characterID *int) {
	var universal, structures []int
	for _, id := range missing {
		if id >= structureIDFloor {
			structures = append(structures, id)
		} else {
			universal = append(universal, id)
		}
	}
	r.fillUniverseNames(ctx, out, universal)
	if len(structures) > 0 && characterID != nil {
		r.fillStructureNames(ctx, out, structures, *characterID)
	}
}

func (r *Resolver) fillUniverseNames(ctx context.Context, out map[int]string, universal []int) {
	for start := 0; start < len(universal); start += nameBatch {
		end := min(start+nameBatch, len(universal))
		chunk := universal[start:end]
		result, err := r.esi.Post(ctx, "/universe/names", nil, nil, chunk)
		if err != nil {
			log.Printf("bulk name lookup failed for %d ids: %v", len(chunk), err)

			continue
		}
		var entries []store.NameRow
		for _, row := range j.Maps(result) {
			id := j.Int(row["id"])
			name := j.Str(row["name"])
			if id == 0 || name == "" {
				continue
			}
			entries = append(entries, store.NameRow{ID: int64(id), Name: name, Category: j.Str(row["category"])})
			out[id] = name
		}
		if err := r.store.NamePut(ctx, entries); err != nil {
			log.Printf("names: put universe names: %v", err)
		}
	}
}

func (r *Resolver) fillStructureNames(ctx context.Context, out map[int]string, structures []int, characterID int) {
	type box struct {
		id   int
		name string
		ok   bool
	}
	ch := make(chan box, len(structures))
	for _, sid := range structures {
		go func(sid int) {
			name, err := r.structureName(ctx, sid, characterID)
			ch <- box{sid, name, err == nil}
		}(sid)
	}
	var entries []store.NameRow
	for range structures {
		b := <-ch
		if !b.ok {
			continue
		}
		entries = append(entries, store.NameRow{ID: int64(b.id), Name: b.name, Category: "structure"})
		out[b.id] = b.name
	}
	if err := r.store.NamePut(ctx, entries); err != nil {
		log.Printf("names: put structure names: %v", err)
	}
}

func collectNameBuckets(lookup map[string]any, only []string) map[string][]NameMatch {
	onlySet := map[string]struct{}{}
	for _, k := range only {
		onlySet[k] = struct{}{}
	}
	buckets := map[string][]NameMatch{}
	for key, entries := range lookup {
		if len(onlySet) > 0 {
			if _, ok := onlySet[key]; !ok {
				continue
			}
		}
		kind := categoryKind(key)
		if kind == "" {
			kind = key
		}
		for _, entry := range j.Maps(entries) {
			id := j.Int(entry["id"])
			name := j.Str(entry["name"])
			if id == 0 || name == "" {
				continue
			}
			k := strings.ToLower(strings.TrimSpace(name))
			buckets[k] = append(buckets[k], NameMatch{ID: id, Name: name, Category: key, Kind: kind})
		}
	}

	return buckets
}

func preferRank(prefer []string) map[string]int {
	rank := map[string]int{}
	for i, key := range prefer {
		rank[key] = i
	}

	return rank
}

func pickNameResolutions(names []string, buckets map[string][]NameMatch, rank map[string]int) map[string]NameResolution {
	out := map[string]NameResolution{}
	for _, asked := range names {
		wanted := strings.ToLower(strings.TrimSpace(asked))
		matches := append([]NameMatch{}, buckets[wanted]...)
		sort.Slice(matches, lessNameMatch(matches, rank))
		res := NameResolution{Query: strings.TrimSpace(asked)}
		if len(matches) > 0 {
			cp := matches[0]
			res.Chosen = &cp
			res.Alternatives = matches[1:]
		}
		out[wanted] = res
	}

	return out
}

func lessNameMatch(matches []NameMatch, rank map[string]int) func(i, j int) bool {
	return func(i, j int) bool {
		ri, okI := rank[matches[i].Category]
		if !okI {
			ri = len(rank)
		}
		rj, okJ := rank[matches[j].Category]
		if !okJ {
			rj = len(rank)
		}
		if ri != rj {
			return ri < rj
		}
		if matches[i].Category != matches[j].Category {
			return matches[i].Category < matches[j].Category
		}

		return matches[i].ID < matches[j].ID
	}
}

func (r *Resolver) structureName(ctx context.Context, structureID, characterID int) (string, error) {
	result, err := r.esi.Get(ctx, fmt.Sprintf("/universe/structures/%d", structureID), &characterID, nil, nil)
	if err != nil {
		return "", err
	}
	name := j.Str(j.Map(result.Data)["name"])
	if name == "" {
		return fmt.Sprintf("Structure #%d", structureID), nil
	}

	return name, nil
}

func (r *Resolver) blob(ctx context.Context, key string, maxAge *time.Duration) (any, error) {
	raw, err := r.store.BlobGet(ctx, key, maxAge)
	if err != nil || raw == nil {
		return nil, err
	}
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return nil, err
	}

	return v, nil
}

func (r *Resolver) putBlob(ctx context.Context, key string, value any) error {
	raw, err := json.Marshal(value)
	if err != nil {
		return err
	}

	return r.store.BlobPut(ctx, key, raw)
}
